src/vs/editor/contrib/stickyScroll/browser/stickyScroll.ts

```diff
@@ -0,0 +1,338 @@
+		}
+	}
+
+	private _addOutlineRanges(outlineElement: OutlineElement, depth: number) {
+		let currentStartLine: number | undefined = 0;
+		let currentEndLine: number | undefined = 0;
+
+		while (outlineElement) {
+			const kind: SymbolKind = outlineElement.symbol.kind;
+			if (kind === SymbolKind.Class || kind === SymbolKind.Constructor || kind === SymbolKind.Function || kind === SymbolKind.Interface || kind === SymbolKind.Method) {
+				currentStartLine = outlineElement?.symbol.range.startLineNumber as number;
+				currentEndLine = outlineElement?.symbol.range.endLineNumber as number;
+				this._ranges.push([currentStartLine, currentEndLine, depth]);
+				depth--;
+			}
+			if (outlineElement.parent instanceof OutlineElement) {
+				outlineElement = outlineElement.parent;
+			} else {
+				break;
+			}
+		}
+	}
+
+	private async _updateOutlineModel(token: CancellationToken) {
+		if (this._editor.hasModel()) {
+			const model = this._editor.getModel();
+			const outlineModel = await OutlineModel.create(this._languageFeaturesService.documentSymbolProvider, model, token);
+			if (token.isCancellationRequested) {
+				return;
+			}
+			this._ranges = [];
+			for (const outline of outlineModel.children.values()) {
+				if (outline instanceof OutlineElement) {
+					this._findLineRanges(outline, 1);
+				}
+			}
+			this._ranges = this._ranges.sort(function (a, b) {
+				if (a[0] !== b[0]) {
+					return a[0] - b[0];
+				} else if (a[1] !== b[1]) {
+					return b[1] - a[1];
+				} else {
+					return a[2] - b[2];
+				}
+			});
+			let previous: number[] = [];
+			for (const [index, arr] of this._ranges.entries()) {
+				const [start, end, _depth] = arr;
+				if (previous[0] === start && previous[1] === end) {
+					this._ranges.splice(index, 1);
+				} else {
+					previous = arr;
+				}
+			}
+		}
+	}
+
+	private _renderStickyScroll() {
+
+		if (!(this._editor.hasModel())) {
+			return;
+		}
+
+		const lineHeight = this._editor.getOption(EditorOption.lineHeight);
+		const model = this._editor.getModel();
+
+		const scrollTop = this._editor.getScrollTop();
+		let scrollDirection: ScrollDirection;
+		if (this._lastScrollPosition < scrollTop) {
+			scrollDirection = ScrollDirection.Down;
+		} else {
+			scrollDirection = ScrollDirection.Up;
+		}
+		this._lastScrollPosition = scrollTop;
+
+		const scrollTopFromLengthOfArray: number = this._editor.getScrollTop() + this.stickyScrollWidget.codeLineCount * lineHeight + 1;
+		this.stickyScrollWidget.emptyRootNode();
+		const beginningLinesConsidered = new Set();
+
+		for (const [index, arr] of this._ranges.entries()) {
+			const [start, end, depth] = arr;
+			const scrollTopFromDepth: number = this._editor.getScrollTop() + depth * lineHeight + 1;
+
+			if (!beginningLinesConsidered.has(start)) {
+				if (this._editor.getScrollTop() + (depth - 1) * lineHeight + 1 >= (end - 1) * lineHeight && this._editor.getScrollTop() + (depth - 1) * lineHeight < end * lineHeight - 2) {
```

Comment:
It would be good to extract some local variables / constants instead of always writing these. e.g.
```ts
const scrollTop = this._editor.getScrollTop();
const depthDistanceToTop = (depth - 1) * lineHeight; // or better names that reflect what you have in mind
```